controller/src/main/java/io/pravega/controller/server/eventProcessor/CommitEventProcessor.java

```diff
@@ -77,32 +82,75 @@ public CommitEventProcessor(final StreamMetadataStore streamMetadataStore,
-                    } else {
-                        log.debug("Successfully committed transaction {} on stream {}/{}", event.getTxid(),
-                                event.getScope(), event.getStream());
-                        if (processedEvents != null) {
-                            processedEvents.offer(event);
-                        }
-                    }
-                }).join();
+        streamMetadataStore.getActiveEpoch(scope, stream, context, executor).thenComposeAsync(pair -> {
+            List<Integer> segments = pair.getRight();
+            int activeEpoch = pair.getLeft();
+            if (activeEpoch == epoch) {
+                return completeCommit(scope, stream, epoch, segments, txnId, context);
+            } else {
+                return postponeCommitEvent(event);
+            }
+        }).whenCompleteAsync((result, error) -> {
+            if (error != null) {
+                log.error("Failed committing transaction {} on stream {}/{}", txnId, scope, stream);
+            } else {
+                log.debug("Successfully committed transaction {} on stream {}/{}", txnId, scope, stream);
+                if (processedEvents != null) {
+                    processedEvents.offer(event);
+                }
+            }
+        }, executor).join();
     }
 
-    private CompletableFuture<Controller.TxnStatus> notifyCommitToHost(final String scope, final String stream, final int segmentNumber, final UUID txId) {
+    private CompletableFuture<Void> completeCommit(final String scope,
+                                                   final String stream,
+                                                   final int epoch,
+                                                   final List<Integer> segments,
+                                                   final UUID txnId,
+                                                   final OperationContext context) {
+        return notifyCommitToHost(scope, stream, segments, txnId).thenComposeAsync(x ->
+                streamMetadataStore.commitTransaction(scope, stream, epoch, txnId, context, executor), executor)
+                .thenApply(x -> null);
+    }
+
+    private CompletableFuture<Void> postponeCommitEvent(CommitEvent event) {
+        UUID txnId = event.getTxid();
+        log.debug("Transaction {}, pushing back CommitEvent to commitStream", txnId);
+        AckFuture future = this.getSelfWriter().write(event);
+        CompletableFuture<AckFuture> writeComplete = new CompletableFuture<>();
+        future.addListener(() -> writeComplete.complete(future), executor);
+        return writeComplete.thenApplyAsync(ackFuture -> {
+            try {
+                // ackFuture is complete by now, so we can do a get without blocking
+                ackFuture.get();
+                log.debug("Transaction {}, sent request to commitStream", txnId);
+                return null;
+            } catch (InterruptedException e) {
```

Comment:
we should have a good story around what happens if we are not able to write the event back into the stream.
in my PR i retry indefinitely until we are able to write the event.
any exception thrown in ackFuture will right now result in a failure which doesnt even lead to a retry and we can have commit on the transaction lying in limbo as the event is lost.